0023-keymanager-secret-sharing.md

```diff
@@ -0,0 +1,694 @@
+  in the master secret protocol).
+
+### Example apps
+
+Current and future applications:
+
+- Master secrets (for generation and replication master secrets)
+
+- Ephemeral secrets (for generation and replication of ephemeral secrets)
+
+- CPU change (for detecting wether the CPU has changed)
+
+- CHURP (secret sharing scheme)
+
+- Key derivation center (secret sharing scheme)
+
+### App trait
+
+Every application should implement the following trait.
+
+```rust
+/// Key manager application.
+pub trait App {
+  /// Register RPC methods to the dispatcher.
+  ///
+  /// Registered methods are exposed via enclave RPC to the local host
+  /// or remote clients.
+  fn register(&self, ...);
+
+  /// Initialize the application on startup.
+  ///
+  /// Use runtime ID, untrusted local storage, host data, consensus verifier,
+  /// runtime identity, etc. to start the application.
+  fn init(&self, ...);
+}
+```
+
+Each application should register RPC methods and adhere to the naming
+convention `app.Method`.
+
+#### Example 1
+
+```rust
+/// Master secrets key manager application.
+pub trait MasterSecrets {
+  fn generate(&self);
+  fn load(&self);
+  fn replicate(&self);
+  fn key_pair(&self);
+  fn private_key(&self);
+  fn public_key(&self);
+  fn symmetric_key(&self);
+  fn update_status(&self);
+}
+```
+
+Methods:
+
+- `MasterSecrets.Generate`
+
+- `MasterSecrets.Load`
+
+- `MasterSecrets.Replicate`
+
+- `MasterSecrets.KeyPair`
+
+- `MasterSecrets.PrivateKey`
+
+- `MasterSecrets.PublicKey`
+
+- `MasterSecrets.SymmetricKey`
+
+- `MasterSecrets.UpdateStatus`
+
+#### Example 2
+
+```rust
+/// CPU change detection key manager application.
+pub trait CPUChangeDetection {
+  fn encrypt(&self);
+  fn decrypt(&self);
+}
+```
+
+Methods:
+
+- `CpuChange.Encrypt`
+
+- `CpuChange.Decrypt`
+
+### App worker
+
+Each key manager application should have a dedicated worker on the host node
+responsible for communicating with the app and ensuring its consensus view
+is up-to-date.
+
+```go
+// AppWorker is a key manager application worker.
+type AppWorker interface {
+  // Start starts the worker.
+  Start()
+  // Stop stops the worker.
+  Stop()
+}
+```
+
+#### Example
+
+A master secret worker should be responsible for participating in the master
+secret protocol.
+
+## CHURP
+
+CHURP is a proactive secret sharing scheme in which the committee of nodes
+storing a secret can change over time.
+
+### Requirements
+
+- Support multiple secrets.
+
+  - Each group of clients can share a unique secret.
+
+  - Secrets may vary in terms of committee size, handoff intervals,
+    security levels, etc.
+
+  - State re-encryption should be possible with keys derived from
+    a new shared secret.
+
+- Key manager restarts should not lose shares or committed data.
+
+### Protocol
+
+#### Offline
+
+- Assign non-zero ID numbers to all nodes and associate them with their
+  public keys.
+
+- The key manager owner selects a Schnorr group based on security
```

Comment:
For ECC-based KDC, we wouldn't use a Schnorr prime.  The prime field version is significantly slower than the ECC version, so I'd recommend going the ECC route.

Reply:
Will generalize so that we can support different groups.

Reply:
the difference between ECC-KDC and prime field KDC is pretty large based on the benchmarks that i've done.  for reconstructing a previously seen derived key w/ possibly compromised shareholders (key-share managers), ecc is a few tens of milliseconds, and prime field is about a second of CPU time.  for reconstructing a new derived key (e.g., for a new contract), it's a few tens x that time, which makes non-ecc probably impractical.  there are still opportunities for performance improvements, so it may still be possible to bring prime field KDC w/ corrupt shares into the realm of usefulness.

CHURP itself is fast, but i think we want KDC so that the execution nodes are never able to reconstruct the shared master secret.